src/transaction/log_recovery_redo_parallel.cpp

```diff
@@ -752,16 +756,122 @@ namespace cublog
-				       LOG_ZIP &undo_unzip_support, LOG_ZIP &redo_unzip_support)
+  redo_job_impl::read_record_and_redo (THREAD_ENTRY *thread_p, log_reader &log_pgptr_reader, const VPID &rcv_vpid,
+				       const log_lsa &rcv_lsa, LOG_ZIP &undo_unzip_support, LOG_ZIP &redo_unzip_support)
   {
     log_pgptr_reader.advance_when_does_not_fit (sizeof (T));
     const T log_rec = log_pgptr_reader.reinterpret_copy_and_add_align<T> ();
     log_rv_redo_record_sync<T> (thread_p, log_pgptr_reader, log_rec, rcv_vpid, rcv_lsa,
 				m_end_redo_lsa, m_log_rtype,
 				undo_unzip_support, redo_unzip_support);
   }
+
+  /*********************************************************************
+   * reusable_jobs_stack - definition
+   *********************************************************************/
+
+  reusable_jobs_stack::reusable_jobs_stack ()
+    : m_stack_size { 0 }
+  {
+  }
+
+  void reusable_jobs_stack::initialize (std::size_t a_stack_size, const log_lsa *a_end_redo_lsa,
+					bool force_each_page_fetch)
+  {
+    assert (m_pop_stack.empty ());
+    assert (m_push_stack.empty ());
+
+    m_stack_size = a_stack_size;
+    for (std::size_t idx = 0; idx < m_stack_size; ++idx)
+      {
+	redo_parallel::redo_job_base *job = new redo_job_impl (a_end_redo_lsa, force_each_page_fetch, this);
+	m_pop_stack.push (job);
+      }
+  }
+
+  void reusable_jobs_stack::initialize (std::size_t a_stack_size,
+					std::function<redo_parallel::redo_job_base * ()> a_job_factory)
+  {
+    assert (m_pop_stack.empty ());
+    assert (m_push_stack.empty ());
+
+    m_stack_size = a_stack_size;
+    for (std::size_t idx = 0; idx < m_stack_size; ++idx)
+      {
+	redo_parallel::redo_job_base *job = a_job_factory ();
+	m_pop_stack.push (job);
+      }
+  }
+
+  reusable_jobs_stack::~reusable_jobs_stack ()
+  {
+    assert ((m_pop_stack.size () + m_push_stack.size ()) == m_stack_size);
+
+    while (!m_push_stack.empty ())
+      {
+	redo_parallel::redo_job_base *const push_job = m_push_stack.top ();
+	m_push_stack.pop ();
+	delete push_job;
+      }
+    while (!m_pop_stack.empty ())
+      {
+	redo_parallel::redo_job_base *const pop_job = m_pop_stack.top ();
+	m_pop_stack.pop ();
+	delete pop_job;
+      }
+  }
+
+  redo_parallel::redo_job_base *reusable_jobs_stack::blocking_pop ()
+  {
+    if (!m_pop_stack.empty ())
+      {
+	redo_parallel::redo_job_base *const pop_job = m_pop_stack.top ();
+	m_pop_stack.pop ();
+	return pop_job;
+      }
+    else
+      {
+	while (true)
+	  {
+	    {
+	      std::lock_guard<std::mutex> stack_lockg { m_mutex };
+	      m_pop_stack.swap (m_push_stack);
+	    }
+	    if (!m_pop_stack.empty ())
+	      {
+		redo_parallel::redo_job_base *const pop_job = m_pop_stack.top ();
+		m_pop_stack.pop ();
+		return pop_job;
+	      }
+	    else
+	      {
+		std::unique_lock<std::mutex> ulock { m_mutex };
+		m_jobs_available_on_push_stack_cv.wait (ulock, [this] ()
+		{
+		  return !m_push_stack.empty ();
+		});
+	      }
```

Comment:
The code becomes shorter if conditional wait for push stack not being empty is checked from the beginning. If push stack is already not empty, the conditional wait will return immediately. If the push stack is empty, it waits until it is not empty and the stacks are not swapped unnecessarily.
```suggestion
	    std::unique_lock<std::mutex> ulock { m_mutex };
	    m_jobs_available_on_push_stack_cv.wait (ulock, [this] ()
	    {
		return !m_push_stack.empty ();
	    });
	    m_pop_stack.swap (m_push_stack);
	    assert (!m_pop_stack.empty ());
	    redo_parallel::redo_job_base *const pop_job = m_pop_stack.top ();
	    m_pop_stack.pop ();
	    return pop_job;
```

EDIT: The while loop also becomes unnecessary.

Reply:
Done